### PR TITLE
Update xml comments for ServiceClient

### PR DESCRIPTION
**Pull request by abhipsaMisra**
Fix for #1808

### CODE REVIEW

File: iothub/service/tests/ServiceClientTests.cs [hunk 1]
```diff
@@ -19,23 +19,6 @@ namespace Microsoft.Azure.Devices.Api.Test
...
-        public async Task PurgeMessageQueueTest()
-        {
-            // Arrange Moq
-            Tuple<Mock<IHttpClientHelper>, AmqpServiceClient, PurgeMessageQueueResult> setupParameters = this.SetupPurgeMessageQueueTests();
-            Mock<IHttpClientHelper> restOpMock = setupParameters.Item1;
-            AmqpServiceClient serviceClient = setupParameters.Item2;
-            PurgeMessageQueueResult expectedResult = setupParameters.Item3;
-
-            // Execute method under test
-            PurgeMessageQueueResult result = await serviceClient.PurgeMessageQueueAsync("TestDevice").ConfigureAwait(false);
```
Comment (abhipsaMisra) [inline iothub/service/tests/ServiceClientTests.cs:32]: This test was against the internal service client implementation

File: iothub/service/src/ServiceClient.cs [hunk 2]
```diff
@@ -99,111 +100,74 @@ public static ServiceClient CreateFromConnectionString(string connectionString,
...
         /// <summary>
-        /// Removes all messages from a device's queue.
+        /// Removes all cloud to device messages from a device's queue.
         /// </summary>
-        /// <param name="deviceId"></param>
-        /// <returns></returns>
-        public abstract Task<PurgeMessageQueueResult> PurgeMessageQueueAsync(string deviceId);
+        /// <param name="deviceId">The device identifier for the target device.</param>
+        /// <param name="cancellationToken">A cancellation token to cancel the operation.</param>
+        public abstract Task<PurgeMessageQueueResult> PurgeMessageQueueAsync(string deviceId, CancellationToken cancellationToken = default);
```
Comment (abhipsaMisra) [inline iothub/service/src/ServiceClient.cs:125]: CancellationToken cancellationToken = default is the same as CancellationToken.None

File: iothub/service/src/AmqpServiceClient.cs [hunk 3]
```diff
@@ -202,12 +196,6 @@ public override FileNotificationReceiver<FileNotification> GetFileNotificationRe
             return _fileNotificationReceiver;
         }
 
-        // This call is executed over HTTP.
-        public override Task<ServiceStatistics> GetServiceStatisticsAsync()
```
Comment (timtay-microsoft) [inline iothub/service/src/AmqpServiceClient.cs:206]: Why are these methods being deleted? Is there a base implementation somewhere in the inheritance of this class that defines this exact implementation already?

Comment (abhipsaMisra) [reply iothub/service/src/AmqpServiceClient.cs:206]: These methods (1) call the method (2) that takes in a Cancellation token, but by simply passing in CancellationToken.None as the token value. That is the same as calling (2) without a cancellation token, since those methods declare CancellationToken to be an optional param.

File: iothub/service/src/AmqpServiceClient.cs [hunk 4]
```diff
@@ -159,12 +159,6 @@ public async override Task SendAsync(string deviceId, Message message, TimeSpan?
             }
         }
 
-        // This call is executed over HTTP.
```
Comment (bikamani) [inline iothub/service/src/AmqpServiceClient.cs:162]: What are the implications of deleting these methods? How is it being referenced?

Comment (abhipsaMisra) [reply iothub/service/src/AmqpServiceClient.cs:162]: No implications of deleting these, the same functionality is still available through the overload which takes in an optional cancellation token

**Review by timtay-microsoft — APPROVED** [review]

File: iothub/service/src/AmqpServiceClient.cs (hunk 4 above)
Comment (bikamani) [reply iothub/service/src/AmqpServiceClient.cs:162]: Got it.

File: iothub/service/src/ServiceClient.cs [hunk 5]
```diff
@@ -99,111 +100,74 @@ public static ServiceClient CreateFromConnectionString(string connectionString,
...
-        /// Send a one-way notification to the specified device
+        /// Send a cloud to device message to the specified device.
         /// </summary>
-        /// <param name="deviceId">The device identifier for the target device</param>
-        /// <param name="message">The message containing the notification</param>
-        /// <param name="timeout">The operation timeout override. If not used uses OperationTimeout default</param>
-        /// <returns></returns>
+        /// <param name="deviceId">The device identifier for the target device.</param>
+        /// <param name="message">The cloud to device message.</param>
+        /// <param name="timeout">The operation timeout which defaults to 1 minute, if unspecified.</param>
```
Comment (bikamani) [inline iothub/service/src/ServiceClient.cs:117]: 👍

File: iothub/service/src/ServiceClient.cs [hunk 6]
```diff
@@ -99,111 +100,74 @@ public static ServiceClient CreateFromConnectionString(string connectionString,
...
         /// </summary>
         /// <returns>An instance of the FeedbackReceiver</returns>
         public abstract FeedbackReceiver<FeedbackBatch> GetFeedbackReceiver();
 
         /// <summary>
-        /// Get the FeedbackReceiver
+        /// Get the FileNotificationReceiver
         /// </summary>
-        /// <returns>An instance of the FeedbackReceiver</returns>
+        /// <returns>An instance of the FileNotificationReceiver</returns>
```
Comment (bikamani) [inline iothub/service/src/ServiceClient.cs:136]: I do not have any suggestions but do you think it would be good to add some more description around this explaining FileNotificationReceiver?

File: iothub/service/src/ServiceClient.cs [hunk 7]
```diff
@@ -99,111 +100,74 @@ public static ServiceClient CreateFromConnectionString(string connectionString,
...
-        public abstract Task<CloudToDeviceMethodResult> InvokeDeviceMethodAsync(string deviceId, string moduleId, CloudToDeviceMethod cloudToDeviceMethod);
+        /// <param name="deviceId">The device identifier for the target device.</param>
+        /// <param name="cloudToDeviceMethod">Device method parameters (passthrough to the device).</param>
+        /// <param name="cancellationToken">A cancellation token to cancel the operation.</param>
+        /// <returns>Method result.</returns>
+        public abstract Task<CloudToDeviceMethodResult> InvokeDeviceMethodAsync(string deviceId, CloudToDeviceMethod cloudToDeviceMethod, CancellationToken cancellationToken = default);
 
         /// <summary>
-        /// Interactively invokes a method on device
+        /// Interactively invokes a method on device.
```
Comment (bikamani) [inline iothub/service/src/ServiceClient.cs:156]: Is this on a device or module?

File: iothub/service/src/AmqpServiceClient.cs [hunk 8]
```diff
@@ -159,12 +159,6 @@ public async override Task SendAsync(string deviceId, Message message, TimeSpan?
...
-        public override Task<PurgeMessageQueueResult> PurgeMessageQueueAsync(string deviceId)
-        {
-            return PurgeMessageQueueAsync(deviceId, CancellationToken.None);
-        }
-
         // This call is executed over HTTP.
```
Comment (drwill-ms) [inline iothub/service/src/AmqpServiceClient.cs:162]: Why // instead of /// doc comment?

File: iothub/service/src/ServiceClient.cs [hunk 9]
```diff
@@ -29,7 +29,8 @@ public enum TransportType
 #pragma warning restore CA1707 // Identifiers should not contain underscores
 
     /// <summary>
-    /// Contains methods that services can use to send messages to devices
+    /// Contains methods that services can use to send messages to devices/ modules,
```
Comment (drwill-ms) [inline iothub/service/src/ServiceClient.cs:32]: devices/modules (no space)

File: iothub/service/src/ServiceClient.cs [hunk 10]
```diff
@@ -29,7 +29,8 @@ public enum TransportType
...
+    /// invoke a direct method on a device/ module and receive file upload and cloud to device message delivery notifications.
```
Comment (drwill-ms) [inline iothub/service/src/ServiceClient.cs:33]: This reads like a run-on sentence until one realizes that file upload and C2D are a compound adjective for notifications. I'd reword it to be like "and delivery notifications for file upload and cloud-to-device operations."

File: iothub/service/src/ServiceClient.cs [hunk 11]
```diff
@@ -99,111 +100,74 @@ public static ServiceClient CreateFromConnectionString(string connectionString,
...
         /// <summary>
-        /// Close the ServiceClient instance
+        /// Close the ServiceClient instance.
         /// </summary>
-        /// <returns></returns>
         public abstract Task CloseAsync();
 
         /// <summary>
-        /// Send a one-way notification to the specified device
+        /// Send a cloud to device message to the specified device.
```
Comment (drwill-ms) [inline iothub/service/src/ServiceClient.cs:113]: cloud-to-device (compound adjective should have dashes)

File: iothub/service/src/ServiceClient.cs [hunk 12]
```diff
@@ -42,11 +43,11 @@ internal ServiceClient()
...
 
         /// <summary>
-        /// Create ServiceClient from the specified connection string
+        /// Create an instance of ServiceClient from the specified IoT Hub connection string.
         /// </summary>
-        /// <param name="connectionString">Connection string for the IoT Hub</param>
+        /// <param name="connectionString">Connection string for the IoT Hub.</param>
         /// <param name="options">The options that allow configuration of the service client instance during initialization.</param>
-        /// <returns></returns>
+        /// <returns>An instance of ServiceClient.</returns>
```
Comment (vinagesh) [inline iothub/service/src/ServiceClient.cs:50]: use <see cerf="ServiceClient">

File: iothub/service/src/ServiceClient.cs (hunk 10 above)
Comment (vinagesh) [reply iothub/service/src/ServiceClient.cs:33]: cloud-to-device with hyphens?

File: iothub/service/src/ServiceClient.cs [hunk 13]
```diff
@@ -42,11 +43,11 @@ internal ServiceClient()
         }
 
         /// <summary>
-        /// Create ServiceClient from the specified connection string
+        /// Create an instance of ServiceClient from the specified IoT Hub connection string.
         /// </summary>
-        /// <param name="connectionString">Connection string for the IoT Hub</param>
+        /// <param name="connectionString">Connection string for the IoT Hub.</param>
         /// <param name="options">The options that allow configuration of the service client instance during initialization.</param>
```
Comment (vinagesh) [inline iothub/service/src/ServiceClient.cs:49]: I know this already exists but adding <see cref="ServiceClientOptions"> for options might be helpful.

File: iothub/service/src/AmqpServiceClient.cs (hunk 8 above)
Comment (abhipsaMisra) [reply iothub/service/src/AmqpServiceClient.cs:162]: I don't intend for this to be doc comment but rather a code comment to aid developers during debugging

File: iothub/service/src/ServiceClient.cs [hunk 14]
```diff
@@ -66,25 +67,25 @@ public void Dispose()
         protected virtual void Dispose(bool disposing) { }
 
         /// <summary>
-        /// Create ServiceClient from the specified connection string using specified Transport Type
+        /// Create an instance of ServiceClient from the specified IoT Hub connection string using specified Transport Type.
```
Comment (vinagesh) [inline iothub/service/src/ServiceClient.cs:70]: Adding `see<cref="TransportType ">` will be helpful here and below in the transportType param

File: iothub/service/src/ServiceClient.cs [hunk 15]
```diff
@@ -99,111 +100,75 @@ public static ServiceClient CreateFromConnectionString(string connectionString,
...
         /// <summary>
-        /// Send a one-way notification to the specified device
+        /// Send a cloud to device message to the specified device.
         /// </summary>
-        /// <param name="deviceId">The device identifier for the target device</param>
-        /// <param name="message">The message containing the notification</param>
-        /// <param name="timeout">The operation timeout override. If not used uses OperationTimeout default</param>
-        /// <returns></returns>
+        /// <param name="deviceId">The device identifier for the target device.</param>
+        /// <param name="message">The cloud to device message.</param>
```
Comment (vinagesh) [inline iothub/service/src/ServiceClient.cs:116]: same - cloud-to-device message

File: iothub/service/src/ServiceClient.cs [hunk 16]
```diff
@@ -99,111 +100,75 @@ public static ServiceClient CreateFromConnectionString(string connectionString,
...
-        /// <summary>
-        /// Interactively invokes a method on device
-        /// </summary>
-        /// <param name="deviceId">Device Id</param>
-        /// <param name="cloudToDeviceMethod">Device method parameters (passthrough to device)</param>
-        /// <param name="cancellationToken">Cancellation Token</param>
-        /// <returns>Method result</returns>
-        public abstract Task<CloudToDeviceMethodResult> InvokeDeviceMethodAsync(string deviceId, CloudToDeviceMethod cloudToDeviceMethod, CancellationToken cancellationToken);
+        /// <param name="cancellationToken">A cancellation token to cancel the operation.</param>
+        /// <returns>The service statistics that can be retrieved from IotHub, eg. the number of device connected to the hub.</returns>
```
Comment (vinagesh) [inline iothub/service/src/ServiceClient.cs:143]: IoT hub

Comment (vinagesh) [reply iothub/service/src/ServiceClient.cs:143]: the number of device(s) - plural

Comment (vinagesh) [reply iothub/service/src/ServiceClient.cs:143]: connection to the IoT hub

File: iothub/service/src/ServiceClient.cs [hunk 17]
```diff
@@ -99,111 +100,75 @@ public static ServiceClient CreateFromConnectionString(string connectionString,
...
-        /// Interactively invokes a method on device
+        /// Interactively invokes a method on a device.
         /// </summary>
-        /// <param name="deviceId">Device Id</param>
-        /// <param name="moduleId">Module Id</param>
-        /// <param name="cloudToDeviceMethod">Device method parameters (passthrough to device)</param>
-        /// <returns>Method result</returns>
-        public abstract Task<CloudToDeviceMethodResult> InvokeDeviceMethodAsync(string deviceId, string moduleId, CloudToDeviceMethod cloudToDeviceMethod);
+        /// <param name="deviceId">The device identifier for the target device.</param>
+        /// <param name="cloudToDeviceMethod">Device method parameters (passthrough to the device).</param>
```
Comment (vinagesh) [inline iothub/service/src/ServiceClient.cs:150]: Wondering if we need to say - passthrough to the device. The class itself has a better description I felt - "Parameters to execute a direct method on the device"

File: iothub/service/src/ServiceClient.cs [hunk 18]
```diff
@@ -99,111 +100,75 @@ public static ServiceClient CreateFromConnectionString(string connectionString,
...
+        /// <param name="cancellationToken">A cancellation token to cancel the operation.</param>
+        /// <returns>Method result.</returns>
```
Comment (vinagesh) [inline iothub/service/src/ServiceClient.cs:152]: Adding a <see cref="CloudToDeviceMethodResult"> would be helpful here.

File: iothub/service/src/ServiceClient.cs [hunk 19]
```diff
@@ -99,111 +100,75 @@ public static ServiceClient CreateFromConnectionString(string connectionString,
...
-        /// Removes all messages from a device's queue.
-        /// </summary>
-        /// <param name="deviceId"></param>
-        /// <param name="cancellationToken"></param>
-        /// <returns></returns>
-        public abstract Task<PurgeMessageQueueResult> PurgeMessageQueueAsync(string deviceId, CancellationToken cancellationToken);
-
-        /// <summary>
-        /// Get the FeedbackReceiver
+        /// Get the FeedbackReceiver.
```
Comment (vinagesh) [inline iothub/service/src/ServiceClient.cs:128]: Should we add more comments to describe it?
Feedback messages are delivery acknowledgments for messages sent to a device from IoT Hub.

File: iothub/service/src/ServiceClient.cs [hunk 20]
```diff
@@ -99,111 +100,75 @@ public static ServiceClient CreateFromConnectionString(string connectionString,
...
         /// </summary>
         /// <returns>An instance of the FeedbackReceiver</returns>
```
Comment (vinagesh) [inline iothub/service/src/ServiceClient.cs:130]: cref should be helpful here.

File: iothub/service/src/ServiceClient.cs [hunk 21]
```diff
@@ -99,111 +100,75 @@ public static ServiceClient CreateFromConnectionString(string connectionString,
...
         /// </summary>
-        /// <param name="deviceId">Device Id</param>
-        /// <param name="moduleId">Module Id</param>
-        /// <param name="cloudToDeviceMethod">Device method parameters (passthrough to device)</param>
-        /// <param name="cancellationToken">Cancellation Token</param>
-        /// <returns>Method result</returns>
-        public abstract Task<CloudToDeviceMethodResult> InvokeDeviceMethodAsync(string deviceId, string moduleId, CloudToDeviceMethod cloudToDeviceMethod, CancellationToken cancellationToken);
+        /// <param name="deviceId">The device identifier for the target device.</param>
+        /// <param name="moduleId">The module identifier for the target module.</param>
+        /// <param name="cloudToDeviceMethod">Method parameters (passthrough to the module).</param>
```
Comment (vinagesh) [inline iothub/service/src/ServiceClient.cs:160]: Wondering if we need to say - passthrough to the device. Is this better "Parameters to execute a direct method on the module"?

File: iothub/service/src/ServiceClient.cs [hunk 22]
```diff
@@ -99,111 +100,75 @@ public static ServiceClient CreateFromConnectionString(string connectionString,
...
+        /// <param name="cancellationToken">A cancellation token to cancel the operation.</param>
+        /// <returns>Method result.</returns>
```
Comment (vinagesh) [inline iothub/service/src/ServiceClient.cs:162]: Adding a `<see cref="CloudToDeviceMethodResult">` would be helpful here.

File: iothub/service/src/ServiceClient.cs [hunk 23]
```diff
@@ -99,111 +100,75 @@ public static ServiceClient CreateFromConnectionString(string connectionString,
...
+        public abstract Task<CloudToDeviceMethodResult> InvokeDeviceMethodAsync(string deviceId, string moduleId, CloudToDeviceMethod cloudToDeviceMethod, CancellationToken cancellationToken = default);
 
         /// <summary>
-        /// Send a one-way notification to the specified device module
+        /// Send a cloud to device message to the specified module.
```
Comment (vinagesh) [inline iothub/service/src/ServiceClient.cs:166]: cloud-to-device here and below

File: iothub/service/src/ServiceClient.cs [hunk 24]
```diff
@@ -99,111 +100,75 @@ public static ServiceClient CreateFromConnectionString(string connectionString,
...
         /// </summary>
-        /// <param name="deviceId">The device identifier for the target device</param>
-        /// <param name="moduleId">The module identifier for the target device module</param>
-        /// <param name="message">The message containing the notification</param>
-        /// <returns></returns>
-        public abstract Task SendAsync(string deviceId, string moduleId, Message message);
+        /// <param name="deviceId">The device identifier for the target device.</param>
+        /// <param name="moduleId">The module identifier for the target module.</param>
+        /// <param name="message">The cloud to module message.</param>
```
Comment (vinagesh) [inline iothub/service/src/ServiceClient.cs:170]: cloud-to-module

File: iothub/service/src/AmqpServiceClient.cs (hunk 8 above)
Comment (drwill-ms) [reply iothub/service/src/AmqpServiceClient.cs:162]: The benefit of doc comments is they'll show up in a tooltip. Not interested in that?

File: iothub/service/src/ServiceClient.cs (hunk 10 above)
Comment (drwill-ms) [reply iothub/service/src/ServiceClient.cs:33]: yes because it is a compound adjective

File: iothub/service/src/ServiceClient.cs (hunk 16 above)
Comment (bikamani) [reply iothub/service/src/ServiceClient.cs:143]: or IoT Hub?

File: iothub/service/src/ServiceClient.cs (hunk 19 above)
Comment (drwill-ms) [reply iothub/service/src/ServiceClient.cs:128]: Yes, I'd add more context. Your second sentence is pretty good.

File: iothub/service/src/ServiceClient.cs [hunk 25]
```diff
@@ -99,111 +100,75 @@ public static ServiceClient CreateFromConnectionString(string connectionString,
...
         /// <summary>
-        /// Get the FeedbackReceiver
+        /// Get the FileNotificationReceiver
         /// </summary>
-        /// <returns>An instance of the FeedbackReceiver</returns>
+        /// <returns>An instance of the FileNotificationReceiver</returns>
         public abstract FileNotificationReceiver<FileNotification> GetFileNotificationReceiver();
 
         /// <summary>
         /// Gets service statistics for the Iot Hub.
```
Comment (drwill-ms) [inline iothub/service/src/ServiceClient.cs:140]: IoT (capital T)

Obviously not when it is a variable, but when it is normal English we should case our acronyms appropriately.

**Review by drwill-ms — APPROVED** [review]

File: iothub/service/src/ServiceClient.cs (hunk 12 above)
Comment (abhipsaMisra) [reply iothub/service/src/ServiceClient.cs:50]: I remember receiving feedback in previous PRs that linking the same type in a doc comment for the type was redundant; i.e. linking ServiceClient in doc comments for ServiceClient.
Let me know if you feel otherwise.

File: iothub/service/src/ServiceClient.cs (hunk 17 above)
Comment (bikamani) [inline iothub/service/src/ServiceClient.cs:150]: I have always wondered why we document as "device method" and not "direct method" which is what MS Docs talk use?

File: iothub/service/src/ServiceClient.cs (hunk 12 above)
Comment (vinagesh) [reply iothub/service/src/ServiceClient.cs:50]: I see we follow this pattern everywhere, so pointed it out. This could be a good topic for parking lot. We do the same in track 2 repos as well so would be good be consistent.

File: iothub/service/src/ServiceClient.cs (hunk 6 above)
Comment (vinagesh) [reply iothub/service/src/ServiceClient.cs:136]: +1 to more description

File: iothub/service/src/ServiceClient.cs (hunk 16 above)
Comment (vinagesh) [reply iothub/service/src/ServiceClient.cs:143]: From what I've heard in the past, we use IoT hub.

Comment (bikamani) [reply iothub/service/src/ServiceClient.cs:143]: Interesting, would love to know the reasoning. 
I see IoT **H**ub in many other places in comments in this PR, would like to see it consistent whatever we decide.

**Review by vinagesh — APPROVED** [review]

**Review by bikamani — APPROVED** [review]
small changes as per commented.

File: iothub/service/src/ServiceClient.cs (hunk 16 above)
Comment (vinagesh) [reply iothub/service/src/ServiceClient.cs:143]: I know it is inconsistent but from previous PRs feedbacks, I understand that IoT hub is what makes sense. I guess we should decide as a team which to pick. I'm not an expert in this stuff. So would not mind some other opinions.

File: iothub/service/src/AmqpServiceClient.cs (hunk 8 above)
Comment (abhipsaMisra) [reply iothub/service/src/AmqpServiceClient.cs:162]: In this case, we have this nicely documented in our readme here, so this is mostly for us: https://github.com/Azure/azure-iot-sdk-csharp#iot-hub-service-sdk
